instrumentation/opentelemetry_ecto/test/opentelemetry_ecto_test.exs

```diff
@@ -122,6 +125,65 @@ defmodule OpentelemetryEctoTest do
   end
 
+  test "preloads in sequence are tied to the parent span" do
+    user = Repo.insert!(%User{email: "opentelemetry@erlang.org"})
+    Repo.insert!(%Post{body: "We got traced!", user: user})
+    Repo.insert!(%Comment{body: "We got traced!", user: user})
+
+    attach_handler()
+
+    Tracer.with_span "parent span" do
+      Repo.all(Query.from(User, preload: [:posts, :comments]), in_parallel: false)
+    end
+
+    assert_receive {:span, span(span_id: root_span_id, name: "parent span")}
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:users")}
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:posts")}
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:comments")}
+  end
+
+  test "preloads in parallel are tied to the parent span" do
+    user = Repo.insert!(%User{email: "opentelemetry@erlang.org"})
+    Repo.insert!(%Post{body: "We got traced!", user: user})
+    Repo.insert!(%Comment{body: "We got traced!", user: user})
+
+    attach_handler()
+
+    Tracer.with_span "parent span" do
+      Repo.all(Query.from(User, preload: [:posts, :comments]))
+    end
+
+    assert_receive {:span, span(span_id: root_span_id, name: "parent span")}
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:users")}
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:posts")}
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:comments")}
+  end
+
+  test "nested query preloads are tied to the parent span" do
+    user = Repo.insert!(%User{email: "opentelemetry@erlang.org"})
+    Repo.insert!(%Post{body: "We got traced!", user: user})
+    Repo.insert!(%Comment{body: "We got traced!", user: user})
+
+    attach_handler()
+
+    Tracer.with_span "parent span" do
+      users_query = from u in User, preload: [:posts, :comments]
+      comments_query = from c in Comment, preload: [user: ^users_query]
+      Repo.all(Query.from(User, preload: [:posts, comments: ^comments_query]))
+    end
+
+    assert_receive {:span, span(span_id: root_span_id, name: "parent span")}
+    # root query
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:users")}
+    # comments preload
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:comments")}
+    # users preload
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:users")}
+    # preloads of user
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:posts")}
+    assert_receive {:span, span(parent_span_id: ^root_span_id, name: "opentelemetry_ecto.test_repo.query:comments")}
```

Comment:
The result here should not be every span having the root span of the overall query as they are nested. If we examine the pdict throughout this test, we can see how the last few queries are spawned by one of the spawned tasks (multiple caller depth).

```
## this query (in a task) spawns the next two, therefore it should be their parent

"#PID<0.434.0> -- SELECT u0.\"id\", u0.\"email\", u0.\"id\" FROM \"users\" AS u0 WHERE (u0.\"id\" = $1)"
{:dictionary,
 [
   "$ancestors": [#PID<0.432.0>],
   ...
   "$callers": [#PID<0.415.0>],
  ...
 ]}

## this is a spawned task from <0.434.0>

21:06:23.496 [debug] QUERY OK source="users" db=0.7ms queue=0.9ms
SELECT u0."id", u0."email", u0."id" FROM "users" AS u0 WHERE (u0."id" = $1) [1]
"#PID<0.437.0> -- SELECT p0.\"id\", p0.\"body\", p0.\"user_id\", p0.\"user_id\" FROM \"posts\" AS p0 WHERE (p0.\"user_id\" = $1) ORDER BY p0.\"user_id\""
{:dictionary,
 [
   "$ancestors": [#PID<0.436.0>],
   "$callers": [#PID<0.434.0>, #PID<0.415.0>],
   "$initial_call": {:erlang, :apply, 2}
 ]}

21:06:23.497 [debug] QUERY OK source="posts" db=0.7ms
SELECT p0."id", p0."body", p0."user_id", p0."user_id" FROM "posts" AS p0 WHERE (p0."user_id" = $1) ORDER BY p0."user_id" [1]

## this is also a spawned task from <0.434.0>

"#PID<0.438.0> -- SELECT c0.\"id\", c0.\"body\", c0.\"user_id\", c0.\"user_id\" FROM \"comments\" AS c0 WHERE (c0.\"user_id\" = $1) ORDER BY c0.\"user_id\""
{:dictionary,
 [
   "$ancestors": [#PID<0.436.0>],
   "$callers": [#PID<0.434.0>, #PID<0.415.0>],
   "$initial_call": {:erlang, :apply, 2}
 ]}

21:06:23.498 [debug] QUERY OK source="comments" db=0.7ms queue=0.7ms
SELECT c0."id", c0."body", c0."user_id", c0."user_id" FROM "comments" AS c0 WHERE (c0."user_id" = $1) ORDER BY c0."user_id" [1]
```

Reply:
I'm not sure if we can achieve this currently, because I believe Ecto does not emit start/stop events but just a "query" event. So how can we use these spans in between as parent contexts? In the example above, the common ancestor `#PID<0.436.0>` does not have an active span context. And btw, I'm seeing something else when I run the example (just a single entry in `"$callers"` for the preload queries):


```
query: "SELECT u0.\"id\", u0.\"email\" FROM \"users\" AS u0"
#PID<0.415.0>
process info: [
  links: [],
  dictionary: [
    "$__current_otel_ctx": %{
      {:otel_tracer, :span_ctx} => {:span_ctx,
       133037140329346081629534641564206235649, 11340865674359219389, 1, [],
       true, false, true,
       {:otel_span_ets, #Function<2.11792691/1 in :otel_tracer_server.on_end/1>}}
    },
    ...
  ]
]

09:25:41.842 [debug] QUERY OK source="users" db=0.1ms queue=0.2ms
SELECT u0."id", u0."email" FROM "users" AS u0 []
query: "SELECT c0.\"id\", c0.\"body\", c0.\"user_id\", c0.\"user_id\" FROM \"comments\" AS c0 WHERE (c0.\"user_id\" = $1) ORDER BY c0.\"user_id\""
#PID<0.433.0>
process info: [
  links: [#PID<0.431.0>],
  dictionary: [
    "$initial_call": {:erlang, :apply, 2},
    "$ancestors": [#PID<0.431.0>],
    "$callers": [#PID<0.415.0>]
  ]
]
query: "SELECT p0.\"id\", p0.\"body\", p0.\"user_id\", p0.\"user_id\" FROM \"posts\" AS p0 WHERE (p0.\"user_id\" = $1) ORDER BY p0.\"user_id\""
#PID<0.432.0>
process info: [
  links: [#PID<0.431.0>],
  dictionary: [
    "$initial_call": {:erlang, :apply, 2},
    "$ancestors": [#PID<0.431.0>],
    "$callers": [#PID<0.415.0>]
  ]
]
```

Reply:
Yeah, it's a good question that I don't have a good answer to 😆

I think we'd need some kind of tracker. I'll have to think on that but what you point out makes sense that it isn't going to be achievable at the moment. This is definitely better than nothing but we should probably call it out to users.

Reply:
Hah okay. I've added some docs on this topic. Feel free to change it to your liking 👍